assets/js/atomic/blocks/product-elements/add-to-cart/product-types/variable/variation-attributes/attribute-picker.js

```diff
@@ -19,86 +20,72 @@ import {
  */
 const AttributePicker = ( { attributes, variationAttributes } ) => {
 	const [ variationId, setVariationId ] = useState( 0 );
-
-	// @todo Support default selected attributes in Variation Picker.
-	const [ selectedAttributes, setSelectedAttributes ] = useState( [] );
+	const [ selectedAttributes, setSelectedAttributes ] = useState( {} );
+	const { dispatchActions } = useAddToCartFormContext();
 
 	const attributeNames = Object.keys( attributes );
-	const hasSelectedAttributes =
-		Object.values( selectedAttributes ).filter( Boolean ).length > 0;
 	const hasSelectedAllAttributes =
 		Object.values( selectedAttributes ).filter(
 			( selected ) => selected !== ''
 		).length === attributeNames.length;
```

Comment:
Slightly confused about this comment. The one below questions the need for memo, but if this was momorized it would have similar dependencies. What is your criteria for requiring memo when reviewing?

Reply:
Sorry for the confusion 😄 .  The question here isn't necessarily connected to memoization, but rather if this could be a potentially expensive operation. If the answer is yes, then I think we might want to take a closer look at how we can reduce how often this is calculated (of which memoization might be one option, depending on how often depencies might change.

On first glance, the dependencies here don't look like they'd change as much as some of the dependencies in the below code that is already memoized.

> What is your criteria for requiring memo when reviewing?

Typically I consider a few things:

- memoizing has a cost of it's own with additional overhead. If dependencies change relatively frequently, then any gains from memoization are negated and performance could actually be worse than no memoization.
- does the calculated value get used as a prop in child components or in updating state? If yes, then it's possible that memoization will reduce the amount of times that state update/prop causes component re-renders.
- does the calculated value get used as a dependency for hooks? If yes, then it's possible memoization can reduce the impact on those hooks.
- does the calculated value resolve to a primitive? If so, then memoization is _likely_ not needed (unless that calculation itself is expensive). Otherwise, things that have reference changes (Objects, Arrays, functions) on every call to the component typically do get memoized to ensure child components implementing them don't re-render unnecessarily.

In the case I commented on below, I reviewed the source of `attributes` being fed into the `AttributePicker` component and it looks like they ultimately come from `getAttributes` (`variation-attributes/utils`) and as far as I can tell, that function returns either a new Array or Object (incidentally, I just noticed that, shouldn't it return a consistent type?) everytime it is called. So for the purpose of `attributes` being a dependency on `useMemo` the memoization is not accomplishing anything because the value is rememoized everytime.  So you're actually hurting performance by the use of `useMemo` in this case.